Assets/Scripts/GameController.cs

```diff
@@ -1,10 +1,185 @@
+		get{return _gridHeight;}
+		set{_gridHeight = value;}
+	}
+	// Returns _gridLength
+	public int GridLength {
+		get{return _gridLength;}
+		set{_gridLength = value;}
+	}
+	// Returns _units
+	public List<Unit> Units {
+		get{return _units;}
+		set{_units = value;}
+	}
+	// Returns _selectedToken
+	public static Token SelectedToken {
+		get{return _selectedToken;}
+		set{_selectedToken = value;}
+	}
+	// Returns valid moves array
+	public List<ValidAction> Actions {
+		get{return _actions;}
+		set{_actions = value;}
+	}
+#endregion
+
+
 	void Start() {
-		Grid grid = gameObject.AddComponent<Grid>();
-		grid.CreateGrid(8);
+		// Testing
+		Actions = new List<ValidAction>();
+		Units = new List<Unit>();
+		Test();
+	}
+
+	public void SelectUnit(Unit unit, Token token, bool takenAction) {
+		SelectedToken = token;
+		if(!takenAction) {
+			SetValidActions(unit.name, token);
+		}else{
+			ShowUnitInfo(unit);
+		}
+	}
+
+	public void ShowUnitInfo(Unit unit) {
+
+	}
+
+	public void UnselectUnit() {
+		SelectedToken = null;
+		ClearValidActions();
 	}
 
-}
\ No newline at end of file
+	private void SetValidActions(string unitName, Token token) {
+		float range = 3f; // = Unit.GetRangeFromServer
+		EvalSurroundingTokenActions(unitName, range, token);
+
+		foreach(ValidAction action in Actions) {
+			Tokens[action.col][action.row].SetActionProperties(action.action);
+		}
+	}
+
+	private void EvalSurroundingTokenActions(string unitName, float range, Token token) {
+		Token nextToken;
+		// Start with above token
+		if(token.Y < GridHeight) {
+			nextToken = Tokens[token.X][token.Y+1];
+			EvalNextToken(unitName, range, nextToken);
+		}
+		// Start with left token
+		if(token.X > 0) {
+			nextToken = Tokens[token.X-1][token.Y];
+			EvalNextToken(unitName, range, nextToken);
+		}
+		// Start with below token
+		if(token.Y > 0) {
+			nextToken = Tokens[token.X][token.Y-1];
+			EvalNextToken(unitName, range, nextToken);
+		}
+		// Start with right token
+		if(token.X < GridLength) {
+			nextToken = Tokens[token.X+1][token.Y];
+			EvalNextToken(unitName, range, nextToken);
+		}
+	}
+
+	private void EvalNextToken(string unitName, float range, Token token) {
+		if(!token.CanMove && !token.CanAttack) {
+			float remainingRange;
+			remainingRange = EvalToken(unitName, range, token);
+			if(remainingRange > 0f) {
+				EvalSurroundingTokenActions(unitName, remainingRange, token);
+			}
+		}
+	}
+
+	private float EvalToken(string unitName, float range, Token token) {
+		float remainingRange = range - TerrainMod.TerrainWeight(unitName, token.name);
+		if(remainingRange >= 0f) {
+			Actions.Add(AddValidAction("move", token.X, token.Y));
+		}
+		return (remainingRange >= 0f)? remainingRange : -1f;
+	}
+
+	private void ClearValidActions() {
+		foreach(ValidAction action in Actions) {
+			Tokens[action.col][action.row].SetActionProperties("clear");
+		}
+		Actions.Clear();
+	}
+
+	private ValidAction AddValidAction(string act, int col, int row) {
+		ValidAction action;
```

Comment:
same comments as before / active dev, but in any case, I will change action to valAction